src/curry.ts

```diff
@@ -91,21 +91,51 @@ export const memo = <TFunc extends Function>(
   return memoize({}, func as any, key, ttl) as any as TFunc
 }
 
+export type DebounceFunction<TArgs extends any[]> = {
+  (...args: TArgs): void;
+  /**
+   * Cancels the debounced function
+   */
+  cancel(): void;
+  /**
+   * Runs the debounced function immediately
+   */
+  flush(...args: TArgs): void;
+}
+
 /**
  * Given a delay and a function returns a new function
  * that will only call the source function after delay
- * milliseconds have passed without any invocations
+ * milliseconds have passed without any invocations.
+ *  
+ * The debounce function comes with a `cancel` method 
+ * to cancel delayed `func` invocations and a `flush` 
+ * method to invoke them immediately
  */
 export const debounce = <TArgs extends any[]>(
   { delay }: { delay: number },
   func: (...args: TArgs) => any
-): ((...args: TArgs) => void) => {
-  let timer: any = null
-  const debounced = (...args: TArgs) => {
-    clearTimeout(timer)
-    timer = setTimeout(() => func(...args), delay)
+) => {
+  let timer: NodeJS.Timeout = null
```

Comment:
I used this `NodeJS.Timeout` value because setTimout's own native type (even running a front-end project) is this.

see:

![image](https://user-images.githubusercontent.com/63638196/197858412-3e6a043e-52c5-4098-983b-94a7e3cf7a9c.png)

![image](https://user-images.githubusercontent.com/63638196/197858776-b4b29864-90c3-4887-b34f-de3fce5189e9.png)

![image](https://user-images.githubusercontent.com/63638196/197858894-570cca12-7f5c-4940-8624-5b340d42d125.png)


![image](https://user-images.githubusercontent.com/63638196/197860115-ce5e26ff-c61d-4dd2-953e-91794e5b633f.png)

it seems that even though it is a front end project for type issues, this `node.Timout` is available

I tested it with a project that I used creac react app and it appeared fine

particularly even though it's an internal type that the user can't access, I did it so I don't get any